packages/angular/ssr/src/app-engine.ts

```diff
@@ -0,0 +1,151 @@
+
+  /**
+   * Renders an HTTP request using the appropriate Angular server application and returns a response.
+   *
+   * This method determines the entry point for the Angular server application based on the request URL,
+   * and caches the server application instances for reuse. If the application is in development mode,
+   * the cache is bypassed and a new instance is created for each request.
+   *
+   * If the request URL appears to be for a file (excluding `/index.html`), the method returns `null`.
+   * A request to `https://www.example.com/page/index.html` will render the Angular route
+   * corresponding to `https://www.example.com/page`.
+   *
+   * @param {Request} request - The incoming HTTP request object to be rendered.
+   * @param {unknown} [requestContext] - Optional additional context for the request, such as metadata.
+   * @returns {Promise<Response | null>} A promise that resolves to a Response object or null if the request is file-like.
+   */
+  async render(request: Request, requestContext?: unknown): Promise<Response | null> {
+    // Skip if the request looks like a file but not `/index.html`.
+    const { pathname } = new URL(request.url);
+    if (isFileLike(pathname) && !pathname.endsWith('/index.html')) {
+      return null;
+    }
+
+    const [locale, loadModule] = this.getEntryPointFromUrl(request.url);
+    let serverApp = this.appsCache.get(locale);
+    if (!serverApp || AngularAppEngine.isDevMode) {
+      const { AngularServerApp } = await loadModule();
+
+      serverApp = new AngularServerApp({
+        isDevMode: AngularAppEngine.isDevMode,
+        hooks: this.hooks,
+      });
+
+      this.appsCache.set(locale, serverApp);
+    }
+
+    return serverApp.render(request, requestContext);
+  }
+
+  /**
+   * Retrieves the Angular server application's entry point path and locale based on the requested URL.
+   *
+   * @param {string} url - The URL from which to derive the locale and determine the entry point.
+   * @returns {[string, () => Promise<{ AngularServerApp: typeof AngularServerApp }>]} An array where:
+   * - The first element is the locale derived from the URL.
+   * - The second element is a function that returns a promise resolving to an object containing the `AngularServerApp` type.
+   * @throws {Error} If the entry point path cannot be determined for the given URL.
+   */
+  private getEntryPointFromUrl(url: string): [
+    locale: string,
+    loadModule: () => Promise<{
+      AngularServerApp: typeof AngularServerApp;
+    }>,
+  ] {
+    // Find bundle for locale
+    const entryPoints = this.entryPoints;
+    if (entryPoints.size === 1) {
+      return entryPoints.entries().next().value;
+    }
+
+    const potentialLocale = getPotentialLocaleIdFromUrl(url, this.manifest.basePath);
+    const entryPoint = entryPoints.get(potentialLocale);
+    if (!entryPoint) {
+      throw new Error(`Unable to determine Angular application entrypoint for: ${url}.`);
```

Comment:
> Question: Is this a purely internal error which indicates a bug in Angular or a potential issue with the application (didn't enable the given locale for localization). Should we attempt to fall back to another locale (ex. es-MX -> es) or have a default locale if none matches?

I don't think we should perform any fallbacking. If the website is accessible via http://example.com/es-MX and the user enters http://example.com/es, it should result in a 404 error, as otherwise it could also potentially harmful to SEO.

Note that the locales mentioned here are not the ones available in Angular but the ones the application is specifically built for.

> Should we print the locale we identified and maybe the available locales to help the user understand if they just have a typo or an unsupported locale?

Will update error message.

Reply:
> I don't think we should perform any fallbacking. If the website is accessible via http://example.com/es-MX and the user enters http://example.com/es, it should result in a 404 error, as otherwise it could also potentially harmful to SEO.

I was more thinking of the case when the site supports only `en` and `es` and the user visits `es-MX`. The next best option would be `es` and potentially be usable by those users. If the user visit `fr`, there isn't really a good option either way, so maybe it falls back to the default locale of `en`. That's at least the way most Google products work as I understand it, though the semantics and expectations of the `?hl` parameter is maybe different than Angular where we encode the locale in the URL pathname.

I can see 404 behavior being reasonable too given that the app doesn't support `es-MX`, so asking that for that is clearly incorrect.

Unfortunately I'm mostly monolingual so I don't have a great sense of user or developer's expectations here and what would be the most usable or intuitive behavior. I wonder if there's someone from our localization teams who would have a good sense of the "correct" way of architecting this.

Reply:
I believe localization should be handled like any other resource. If a requested resource is unavailable, it should return a 404 error. We can always adjust this approach later if we decide to change the DX.

Reply:
Is that how things work with the existing SSR toolchain? Maybe the best path forward is to just maintain the current behavior and we can always come back and reevaluate this particular area when we have more information.